controller/controller_server/utils.py

```diff
@@ -14,36 +16,140 @@
+    pools_by_system = parameters.get(config.PARAMETERS_POOLS_BY_SYSTEM)
+    if pools_by_system and secret_uid:
+        return get_pool_by_uid(pools=pools_by_system, uid=secret_uid)
+    return parameters.get(config.PARAMETERS_POOL)
+
+
+def get_volume_id(new_volume, secret_uid):
+    return _get_object_id(new_volume, secret_uid)
 
 
 def get_snapshot_id(new_snapshot):
     return _get_object_id(new_snapshot)
 
 
-def _get_object_id(obj):
+def _get_object_id(obj, secret_uid=None):
+    if secret_uid:
+        return config.PARAMETERS_OBJECT_ID_DELIMITER.join((obj.array_type, secret_uid, obj.id))
     return config.PARAMETERS_OBJECT_ID_DELIMITER.join((obj.array_type, obj.id))
 
 
-def validate_secret(secret):
-    logger.debug("validating secrets")
-    if secret:
-        if not (config.SECRET_USERNAME_PARAMETER in secret and
-                config.SECRET_PASSWORD_PARAMETER in secret and
-                config.SECRET_ARRAY_PARAMETER in secret):
-            raise ValidationException(messages.invalid_secret_message)
+def _validate_secret_uid(uid):
+    if not re.match(config.SECRET_VALIDATION_REGEX, uid.lstrip()):
+        raise ValidationException(messages.invalid_uid_value_message.format(uid))
+    if len(uid) > config.SECRET_UID_MAX_LENGTH:
+        raise ValidationException(
+            messages.parameter_length_is_too_long.format("secret_uid", uid, config.SECRET_UID_MAX_LENGTH))
+
+
+def _validate_secret(secret):
+    if not (config.SECRET_USERNAME_PARAMETER in secret and
+            config.SECRET_PASSWORD_PARAMETER in secret and
+            config.SECRET_ARRAY_PARAMETER in secret):
+        raise ValidationException(messages.invalid_secret_message)
+
+
+def _validate_topologies(secret_topologies):
+    for topologies in secret_topologies:
+        for topology_key, topology_value in topologies.items():
+            split_topology = topology_key.split(config.PARAMETERS_TOPOLOGY_DELIMITER)
+            if len(split_topology) != 2:
+                raise ValidationException(messages.invalid_topology_key_message.format(topology_key))
+            prefix, suffix = split_topology
+            if not re.match(config.SECRET_VALIDATION_REGEX, suffix.lstrip()):
+                raise ValidationException(messages.invalid_topology_key_message.format(suffix))
+            if prefix not in config.TOPOLOGY_PREFIXES:
+                raise ValidationException(messages.invalid_topology_key_message.format(prefix))
+            if not re.match(config.SECRET_VALIDATION_REGEX, topology_value.lstrip()):
+                raise ValidationException(messages.invalid_topology_value_message.format(topology_value))
 
-    else:
-        raise ValidationException(messages.secret_missing_message)
 
+def _validate_secret_config(secret_config_string):
+    secret_config = ast.literal_eval(secret_config_string)
+    for uid, secret_info in secret_config.items():
+        _validate_secret_uid(uid)
+        _validate_secret(secret_info)
```

Comment:
I suggest merging these functions into one. If impossible, please elaborate.

Reply:
`secret` is not bound by `uid`, they are 2 different things, in the topology feature they are connected, without it, I'd have no meaning and does not exist